javalib/src/main/scala/java/util/AbstractList.scala

```diff
@@ -1,3 +1,259 @@
+      i < end
+
+    def next(): E = {
+      last = i
+      i += 1
+      get(last)
+    }
+
+    def previous(): E = {
+      last = i
+      i -= 1
+      get(i)
+    }
+
+    def nextIndex(): Int = i
+
+    def previousIndex(): Int = i - 1
+
+    def add(e: E): Unit = {
+      self.add(i, e)
+      end += 1
+      last = i
+      i += 1
+    }
+
+    def remove(): Unit = {
+      if (i == -1 || i == last)
+        throw new IllegalStateException
+
+      self.remove(last)
+      if (last < i)
+        i -= 1
+      last = i
+      end -= 1
+    }
+
+    def set(e: E): Unit = {
+      if (last == -1)
+        throw new IllegalStateException
+
+      self.set(last, e)
+    }
+  }
+
+  def set(index: Int, element: E): E =
+    throw new UnsupportedOperationException
+
+  def listIterator(index: Int): ListIterator[E] = {
+    checkIndexOnBounds(index)
+    new AbstractListIterator(index, 0, size)
+  }
+
+  def remove(index: Int): E =
+    throw new UnsupportedOperationException
+
+  protected def removeRange(fromIndex: Int, toIndex: Int): Unit = {
+    @tailrec
+    def removeNext(index: Int, iter: ListIterator[E]): Unit = {
+      if (iter.hasNext && index < toIndex) {
+        println(iter.next())
+        iter.remove()
+        removeNext(index + 1, iter)
+      }
+    }
+    removeNext(fromIndex, listIterator(fromIndex))
+  }
+
+  def subList(fromIndex: Int, toIndex: Int): List[E] = {
+    if (fromIndex < 0 || toIndex > size)
+      throw new IndexOutOfBoundsException
+    else if (fromIndex > toIndex)
+      throw new IllegalArgumentException
+
+    class ListView extends AbstractList[E] {
+
+      protected val _fromIndex = fromIndex
+
+      protected var _toIndex = toIndex
+
+      override def add(index: Int, e: E): Unit = {
+        checkIndexOnBoundsOfView(index)
+        self.add(_fromIndex + index, e)
+      }
+
+      override def add(e: E): Boolean = {
+        add(size, e)
+        true
+      }
+
+      override def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
+        checkIndexOnBoundsOfView(index)
+        val success = self.addAll(_fromIndex + index, c)
+        if (success)
+          _toIndex += c.size
+        success
+      }
+
+      override def addAll(c: Collection[_ <: E]): Boolean =
+        addAll(size, c)
+
+      override def clear(): Unit = {
+        self.removeRange(_fromIndex, _toIndex)
+        _toIndex = _fromIndex
+      }
+
+      def get(index: Int): E = {
+        checkIndexInBoundsOfView(index)
+        self.get(_fromIndex + index)
+      }
+
+      override def listIterator(index: Int): java.util.ListIterator[E] = {
+        checkIndexOnBoundsOfView(index)
+
+        // this implementation assumes that the underling iterator is not on a RandomAccess list
+        new ListIterator[E] {
+          val innerIterator = self.listIterator(_fromIndex + index)
+
+          private var i = index
+          private var last = -1
+
+          def hasPrevious(): Boolean = _fromIndex < i
```

Comment:
True, it will be better. Originally it was to subclass with `RandomAccess` and use the index directly on the original array. But then I saw that I change it on the `AbstractListIterator` with more code reuse. I will remove the useless code from this one.